gcp-pubsub/src/main/java/io/micronaut/gcp/pubsub/bind/DefaultSubscriberFactory.java

```diff
@@ -0,0 +1,116 @@
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.cloud.pubsub.v1.SubscriberInterface;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import io.micronaut.context.BeanContext;
+import io.micronaut.gcp.pubsub.configuration.SubscriberConfigurationProperties;
+import io.micronaut.gcp.pubsub.exception.PubSubListenerException;
+import io.micronaut.inject.qualifiers.Qualifiers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Singleton;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Default implementation of {@link SubscriberFactory}.
+ * @author Vinicius Carvalho
+ * @since 2.0.0
+ */
+@Singleton
+public class DefaultSubscriberFactory implements SubscriberFactory, AutoCloseable {
+
+    private final ConcurrentHashMap<ProjectSubscriptionName, Subscriber> subscribers = new ConcurrentHashMap<>();
+    private final TransportChannelProvider transportChannelProvider;
+    private final CredentialsProvider credentialsProvider;
+    private final BeanContext beanContext;
+    private final Collection<SubscriberConfigurationProperties> subscriberConfigurationProperties;
+    private final Logger logger = LoggerFactory.getLogger(DefaultSubscriberFactory.class);
+
+    public DefaultSubscriberFactory(TransportChannelProvider transportChannelProvider,
+                                    CredentialsProvider credentialsProvider,
+                                    BeanContext beanContext,
+                                    Collection<SubscriberConfigurationProperties> subscriberConfigurationProperties
+                                    ) {
+        this.transportChannelProvider = transportChannelProvider;
+        this.credentialsProvider = credentialsProvider;
+        this.beanContext = beanContext;
+        this.subscriberConfigurationProperties = subscriberConfigurationProperties;
+    }
+
+    @Override
+    public Subscriber createSubscriber(SubscriberFactoryConfig config) {
+        Subscriber subscriber = subscribers.compute(config.getSubscriptionName(), (k, v) -> {
+            if (v == null) {
+                Subscriber.Builder builder = Subscriber.newBuilder(config.getSubscriptionName(), config.getReceiver())
+                        .setChannelProvider(this.transportChannelProvider)
+                        .setCredentialsProvider(this.credentialsProvider);
+
+                Optional<SubscriberConfigurationProperties> subscriberConfiguration = subscriberConfigurationProperties.stream().filter(p -> p.getName().equals(config.getSubscriberConfiguration())).findFirst();
+                String executor = subscriberConfiguration.map(s -> s.getExecutor()).orElse(config.getDefaultExecutor());
```

Comment:
I think it might be better to use the context here to retrieve the bean because otherwise the configurations are loaded eagerly, even for those that don't correspond to listeners.

Reply:
@jameskleeh can you elaborate? You mean not inject ` Collection<SubscriberConfigurationProperties> subscriberConfigurationProperties` and use the context to search for the bean? Should I just use `beanContext.getBeansOfType(SubscriberConfigurationProperties.class)` and then find using the filter method for the name? I noticed that if I try `beanContext.getBean(SubscriberConfigurationProperties.class, Qualifiers.byName())` and the bean is not registered I need to deal with an exception. What would be the preferred way in this case then?

Reply:
@viniciusccarvalho There is a `beanContext.findBean` which returns an optional